### PR TITLE
Feature/graceful shutdown

### PR DESCRIPTION
**Pull request by gedge**
### What

added stub healthcheck web server, and graceful shutdown of service

### CODE REVIEW

File: cmd/dp-import-tracker/main.go [hunk 1]
```diff
@@ -107,45 +115,83 @@ func CheckImportJobCompletionState(importAPI *api.ImportAPI, datasetAPI *api.Dat
...
+	tickerChan := make(chan bool)
 	go func() {
 		for range time.Tick(checkForCompleteInstancesTick) {
-			checkForCompletedInstancesChan <- true
+			tickerChan <- true
 		}
 	}()
 
-	for {
+	for looping := true; looping; {
```
Comment (eldeal) [inline cmd/dp-import-tracker/main.go]: If this is just to deal with breaking out of a nested select, naming the loop might be more idiomatic: https://www.goinggo.net/2013/11/label-breaks-in-go.html

File: cmd/dp-import-tracker/main.go [hunk 2]
```diff
@@ -107,45 +115,83 @@ func CheckImportJobCompletionState(importAPI *api.ImportAPI, datasetAPI *api.Dat
...
+				stopTracking := false
+				if err, isFatal := trackedInstances.updateInstanceFromDatasetAPI(ctx, datasetAPI, instanceID); err != nil {
+					log.ErrorC("could not update instance", err, log.Data{"instanceID": instanceID, "isFatal": isFatal})
+					stopTracking = isFatal
+				} else if trackedInstances[instanceID].totalObservations > 0 && trackedInstances[instanceID].observationsInsertedCount >= trackedInstances[instanceID].totalObservations {
+					// the above check on totalObservations ensures that we have updated this instance from the Import API (non-default value)
+					// and that inserts have exceeded the expected
+					logData := log.Data{"instanceID": instanceID, "observations": trackedInstances[instanceID].observationsInsertedCount, "jobID": trackedInstances[instanceID].jobID}
+					log.Debug("import instance complete", logData)
+					log.Error(errors.New("TODO check db for actual count(insertedObservations) now that instance appears to be completed (kafka-double-counting?)"), nil)
```
Comment (eldeal) [inline cmd/dp-import-tracker/main.go]: Don't really think this is needed? just a TODO comment should be fine, seems noisy to log out every time it happens in the mean time (though arguably, the annoyance could encourage us to change it sooner, but I would cynically assume we'd just get used to ignoring the logs)

File: cmd/dp-import-tracker/main.go (hunk 1 above)
Comment (gedge) [reply cmd/dp-import-tracker/main.go]: I originally had a `break` but I actually prefer this.

File: api/dataset_api.go [hunk 3]
```diff
@@ -53,94 +54,108 @@ type JobLinks struct {
...
-	jsonBody, httpCode, err := api.get(path, nil)
+	jsonBody, httpCode, err := api.get(ctx, path, nil)
 	logData["httpCode"] = httpCode
-	if httpCode == http.StatusNotFound {
-		return Instance{}, nil
-	}
 	if err == nil && httpCode != http.StatusOK {
 		err = errors.New("Bad response while getting dataset instance")
+		if httpCode < http.StatusInternalServerError {
+			isFatal = true
```
Comment (daiLlew) [inline api/dataset_api.go]: I think you probably want to log the error was deemed as fatal

File: api/dataset_api.go [hunk 4]
```diff
@@ -53,94 +54,108 @@ type JobLinks struct {
...
+		}
+	} else if err != nil {
```
Comment (daiLlew) [inline api/dataset_api.go]: same here

File: api/dataset_api.go [hunk 5]
```diff
@@ -53,94 +54,108 @@ type JobLinks struct {
...
-func (api *DatasetAPI) GetInstances(vars url.Values) ([]Instance, error) {
+func (api *DatasetAPI) GetInstances(ctx context.Context, vars url.Values) (instances []Instance, err error, isFatal bool) {
 	path := api.url + "/instances"
 	logData := log.Data{"path": path}
-	jsonBody, httpCode, err := api.get(path, vars)
+	jsonBody, httpCode, err := api.get(ctx, path, vars)
 	logData["httpCode"] = httpCode
 	if err == nil && httpCode != http.StatusOK {
 		err = errors.New("Bad response while getting dataset instances")
+		if httpCode < http.StatusInternalServerError {
```
Comment (daiLlew) [inline api/dataset_api.go]: same here

File: api/dataset_api.go [hunk 6]
```diff
@@ -53,94 +54,108 @@ type JobLinks struct {
...
 // GetInstances asks the Dataset API for all instances filtered by vars
-func (api *DatasetAPI) GetInstances(vars url.Values) ([]Instance, error) {
+func (api *DatasetAPI) GetInstances(ctx context.Context, vars url.Values) (instances []Instance, err error, isFatal bool) {
 	path := api.url + "/instances"
 	logData := log.Data{"path": path}
-	jsonBody, httpCode, err := api.get(path, vars)
+	jsonBody, httpCode, err := api.get(ctx, path, vars)
 	logData["httpCode"] = httpCode
 	if err == nil && httpCode != http.StatusOK {
 		err = errors.New("Bad response while getting dataset instances")
```
Comment (daiLlew) [inline api/dataset_api.go]: Probably beyond the scope of this change but I think we need a team discussion on how we handle errors - for example using "github.com/pkg/errors" and making use of `errors.Wrap(...)` and passing the error back up the chain until something deals with it.

File: api/dataset_api.go [hunk 7]
```diff
@@ -53,94 +54,108 @@ type JobLinks struct {
...
+		if httpCode < http.StatusInternalServerError {
+			isFatal = true
+		}
+	} else if err != nil {
+		isFatal = true
```
Comment (daiLlew) [inline api/dataset_api.go]: log fatal error happened

File: api/dataset_api.go [hunk 8]
```diff
@@ -53,94 +54,108 @@ type JobLinks struct {
...
+func (api *DatasetAPI) UpdateInstanceWithNewInserts(ctx context.Context, instanceID string, observationsInserted int32) (err error, isFatal bool) {
 	path := api.url + "/instances/" + instanceID + "/inserted_observations/" + strconv.FormatInt(int64(observationsInserted), 10)
 	logData := log.Data{"url": path}
-	jsonBody, httpCode, err := api.put(path, nil)
+	jsonBody, httpCode, err := api.put(ctx, path, nil)
 	logData["httpCode"] = httpCode
 	logData["jsonBytes"] = jsonBody
 	if err == nil && httpCode != http.StatusOK {
 		err = errors.New("Bad response while updating inserts for job")
+		if httpCode < http.StatusInternalServerError {
```
Comment (daiLlew) [inline api/dataset_api.go]: ditto

File: api/healthcheck.go [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+}
+
+func StopHealthCheck(ctx context.Context) {
+	httpServer.Shutdown(ctx)
+}
+
+// HealthCheck returns the health of the application
+func healthCheck(w http.ResponseWriter, r *http.Request) {
+	log.Debug("Healthcheck endpoint.", nil)
+	w.WriteHeader(http.StatusOK)
```
Comment (daiLlew) [inline api/healthcheck.go]: Add a TODO to implement properly

File: api/dataset_api.go (hunk 3 above)
Comment (gedge) [reply api/dataset_api.go]: it does get logged, a bit later - I've added a comment to that effect

File: api/dataset_api.go (hunk 4 above)
Comment (gedge) [reply api/dataset_api.go]: it does get logged, a bit later - I've added a comment to that effect

File: api/dataset_api.go (hunk 7 above)
Comment (gedge) [reply api/dataset_api.go]: it does get logged, a bit later - I've added a comment to that effect

File: api/dataset_api.go (hunk 8 above)
Comment (gedge) [reply api/dataset_api.go]: I've moved all this error checking to its own function - too much repetition

File: api/healthcheck.go (hunk 9 above)
Comment (gedge) [reply api/healthcheck.go]: done

**Pull request merged**